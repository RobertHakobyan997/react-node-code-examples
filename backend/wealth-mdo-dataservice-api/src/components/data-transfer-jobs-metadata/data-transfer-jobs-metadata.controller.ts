import { Body, Controller, Delete, Get, Param, Post, Put } from '@nestjs/common';
import { ApiBody, ApiOkResponse, ApiParam, ApiTags } from '@nestjs/swagger';
import { PartialType } from '@nestjs/mapped-types';
import { Log } from 'ngpd-merceros-wealth-mdo-common-be/dist/logger/mdo-logger';
import DataTransferJobsMetadata from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/data-transfer-jobs-metadata';

import { DataTransferJobsMetadataService } from './data-transfer-jobs-metadata.service';

export class PartialDataTransferJobsMetadata extends PartialType<DataTransferJobsMetadata>(DataTransferJobsMetadata) {
}

@ApiTags('Data Transfer Job Metadata')
@Controller('data-transfer-jobs-metadata')
export class DataTransferJobsMetadataController {
  constructor(private readonly service: DataTransferJobsMetadataService) {
  }

  @ApiOkResponse({
    description: 'Returns all data transfer jobs metadata',
    type       : [ DataTransferJobsMetadata ],
  })
  @Get()
  @Log('Get all data transfer jobs metadata')
  findAll() {
    return this.service.findAll();
  }

  @ApiOkResponse({
    description: 'Get all data transfer job metadata which active for current GMT time in metadata`s schedule',
    type       : [ DataTransferJobsMetadata ],
  })
  @Get('schedule')
  @Log('Get active data transfer job metadata')
  getSchedule() {
    return this.service.getSchedule();
  }

  @ApiOkResponse({
    description: 'Get list of job keys',
    type       : [ String ]
  })
  @Get('job-keys')
  @Log('Get list of job keys')
  getJobKeys() {
    return this.service.getJobKeys();
  }

  @ApiOkResponse({
    description: 'Returns data transfer job metadata by id',
    type       : [ DataTransferJobsMetadata ],
  })
  @ApiParam({ name: 'id', description: 'Transfer job metadata id', example: '611f68a99f217cce3a4d91f8' })
  @Get(':id')
  @Log('Get data transfer job metadata')
  findById(@Param('id') id: string) {
    return this.service.findById(id);
  }

  @ApiOkResponse({
    description: 'Update data transfer job metadata by id',
    type       : DataTransferJobsMetadata
  })
  @ApiBody({
    description: 'Data transfer job metadata',
    type       : DataTransferJobsMetadata,
  })
  @ApiParam({ name: 'id', description: 'Transfer job metadata id', example: '611f68a99f217cce3a4d91f8' })
  @Put(':id')
  @Log('Update data transfer job metadata')
  update(
    @Param('id') id: string,
    @Body() dataTransferJobsMetadata: PartialDataTransferJobsMetadata,
  ) {
    return this.service.updateOne(id, dataTransferJobsMetadata);
  }

  @ApiOkResponse({
    description: 'Creates data transfer job metadata',
    type       : [ DataTransferJobsMetadata ],
  })
  @ApiBody({ description: 'New data transfer job metadata', type: DataTransferJobsMetadata, isArray: true })
  @Post()
  @Log('Create data transfer job metadata')
  create(
    @Body() dataTransferJobsMetadata: DataTransferJobsMetadata[]
  ) {
    return this.service.create(dataTransferJobsMetadata);
  }

  @ApiOkResponse({
    description: 'Deletes data transfer job metadata by id',
    type       : [ DataTransferJobsMetadata ],
  })
  @ApiParam({ name: 'id', description: 'Transfer job metadata id', example: '611f68a99f217cce3a4d91f8' })
  @Delete(':id')
  @Log('Delete data transfer job metadata')
  delete(@Param('id') id: string) {
    return this.service.delete(id);
  }

  @ApiOkResponse({
    description: 'Returns data transfer job metadata by job key',
    type       : DataTransferJobsMetadata,
  })
  @ApiParam({ name: 'dataTransferJobKey', description: 'Transfer job metadata key', example: 'eagle2Kafka_Entities' })
  @Get('find-by-key/:dataTransferJobKey')
  @Log('Get data transfer job metadata')
  findByJobKey(@Param('dataTransferJobKey') key: string) {
    return this.service.findByJobKey(key);
  }
}
