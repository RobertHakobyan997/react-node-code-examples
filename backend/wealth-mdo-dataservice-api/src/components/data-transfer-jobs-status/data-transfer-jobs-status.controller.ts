import { Body, Controller, Delete, Get, Param, Post, Put } from '@nestjs/common';
import { ApiBody, ApiOkResponse, ApiParam, ApiTags } from '@nestjs/swagger';
import { Log } from 'ngpd-merceros-wealth-mdo-common-be/dist/logger/mdo-logger';
import DataTransferJobsStatus from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/data-transfer-jobs-status';
import { PartialType } from '@nestjs/mapped-types';

import { DataTransferJobsStatusService } from './data-transfer-jobs-status.service';

export class PartialDataTransferJobsStatus extends PartialType<DataTransferJobsStatus>(DataTransferJobsStatus) {
}

@ApiTags('Data Transfer Job Statuses')
@Controller('data-transfer-jobs-status')
export class DataTransferJobsStatusController {
  constructor(private readonly service: DataTransferJobsStatusService) {
  }

  @ApiOkResponse({
    description: 'Returns all data transfer jobs status',
    type       : [ DataTransferJobsStatus ]
  })
  @Get()
  @Log('Get all data transfer jobs status')
  findAll() {
    return this.service.findAll();
  }

  @ApiOkResponse({
    description: 'Returns data transfer job status by id',
    type       : DataTransferJobsStatus,
  })
  @ApiParam({ name: 'id' })
  @Get(':id')
  @Log('Get data transfer job status')
  findById(@Param('id') id: string) {
    return this.service.findById(id);
  }

  @ApiOkResponse({
    description: 'Get files by files metadata id',
    type       : DataTransferJobsStatus,
  })
  @ApiParam({ name: 'jobMetaDataId', description: 'Metadata id' })
  @Get('metadata/:jobMetaDataId')
  @Log('Get files by job metadata id')
  getFilesByMetadataId(@Param('jobMetaDataId') jobMetaDataId: string) {
    return this.service.getJobMetadataStatuses(jobMetaDataId);
  }

  @ApiOkResponse({ description: 'Update data transfer job status by id' })
  @ApiBody({
    description: 'Data transfer job status',
    type       : DataTransferJobsStatus,
  })
  @ApiParam({ name: 'id' })
  @ApiBody({ description: 'Data transfer job status', type: DataTransferJobsStatus })
  @Put(':id')
  @Log('Update data transfer job status of existing item')
  update(
    @Param('id') id: string,
    @Body() dataTransferJobsStatus: PartialDataTransferJobsStatus,
  ) {
    return this.service.update(id, dataTransferJobsStatus);
  }

  @ApiOkResponse({ description: 'Update/Create data transfer job status by job metadata id' })
  @ApiBody({
    description: 'Data transfer job status',
    type       : DataTransferJobsStatus,
  })
  @ApiParam({ name: 'jobMetaDataId', description: 'Metadata id' })
  @ApiBody({ description: 'Data transfer job status', type: DataTransferJobsStatus })
  @Put('metadata/:jobMetaDataId')
  @Log('Update data transfer job status of existing item or create the new one')
  updateJobMetadataStatuses(
    @Param('jobMetaDataId') id: string,
    @Body() dataTransferJobsStatus: Partial<DataTransferJobsStatus>,
  ) {
    return this.service.updateJobMetadataStatuses(id, dataTransferJobsStatus);
  }

  @ApiOkResponse({
    description: 'Creates data transfer job status',
    type       : [ DataTransferJobsStatus ],
  })
  @ApiBody({ description: 'New data transfer job status', type: [ DataTransferJobsStatus ] })
  @Post()
  @Log('Create data transfer job status')
  create(
    @Body() dataTransferJobsStatus: DataTransferJobsStatus[]
  ) {
    return this.service.create(dataTransferJobsStatus);
  }

  @ApiOkResponse({
    description: 'Deletes data transfer job status by id',
    type       : [ DataTransferJobsStatus ],
  })
  @ApiParam({ name: 'id' })
  @Delete(':id')
  @Log('Delete data transfer job status')
  delete(@Param('id') id: string) {
    return this.service.delete(id);
  }
}
