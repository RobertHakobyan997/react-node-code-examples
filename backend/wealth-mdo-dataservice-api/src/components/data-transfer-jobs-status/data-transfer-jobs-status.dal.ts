import { Inject, Injectable } from '@nestjs/common';
import { FilterQuery, Model, UpdateQuery } from 'mongoose';
import { Logger } from 'ngpd-merceros-logger-be';
import { plainToClass } from 'class-transformer';
import DataTransferJobsStatusModel, { DataTransferJobsStatusModelToken } from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/model/mongoose/data-transfer-jobs-status';
import DataTransferJobsStatus from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/data-transfer-jobs-status';
import DataTransferJobsStatusState from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/data-transfer-jobs-status-state';
import { Document } from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/mongodb/document';
import { ObjectId } from 'mongodb';

import { mergeMemo } from '../../core/utils/lodash';

@Injectable()
export class DataTransferJobsStatusDal {
  private readonly logger = new Logger(DataTransferJobsStatusDal.name);

  constructor(@Inject(DataTransferJobsStatusModelToken) private readonly model: Model<Document<DataTransferJobsStatus>>) {
  }

  findAll(filter: FilterQuery<Document<DataTransferJobsStatus>> = {}, sort: string | any = {}, options: any = {}) {
    return this.model
      .find(filter, {}, options)
      .sort(sort)
      .exec();
  }

  findOne(filter: FilterQuery<Document<DataTransferJobsStatus>> = {}, projection: any = {}, options: any = {}) {
    return this.model.findOne(filter, projection, options).exec();
  }

  findById(id: any, options: any = {}) {
    return this.model.findById(id, {}, options).exec();
  }

  insert(docs: any[]) {
    return this.model.insertMany(plainToClass(DataTransferJobsStatusModel, docs));
  }

  async updateByJobMetaDataId(id: any, docs: UpdateQuery<DataTransferJobsStatus>, options: any = {}) {
    const prev = await this.findOne({ jobMetaDataId: new ObjectId(id) });
    const updated = mergeMemo(prev, docs);
    await this.model.findByIdAndUpdate(id, plainToClass(DataTransferJobsStatusModel, updated), options);
    return updated;
  }

  async updateOne(_id: any, docs: UpdateQuery<DataTransferJobsStatus>, options: any = {}) {
    return this.model.updateOne({ _id }, docs, options);
  }

  findByIdAndUpdate(id: any, docs: UpdateQuery<DataTransferJobsStatus>, options: any = {}) {
    return this.model.findByIdAndUpdate(id, plainToClass(DataTransferJobsStatusModel, docs), options);
  }

  findByIdAndDelete(id: any) {
    return this.model.findByIdAndDelete(id);
  }

  update(status: UpdateQuery<DataTransferJobsStatusState>, filter: FilterQuery<Document<DataTransferJobsStatus>> = {}) {
    return this.model.updateOne(filter, plainToClass(DataTransferJobsStatusModel, status));
  }
}
