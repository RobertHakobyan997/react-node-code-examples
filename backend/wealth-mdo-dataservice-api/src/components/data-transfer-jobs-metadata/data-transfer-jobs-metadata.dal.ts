import { Inject, Injectable } from '@nestjs/common';
import { ObjectId } from 'mongodb';
import { FilterQuery, Model, UpdateQuery } from 'mongoose';
import { Logger } from 'ngpd-merceros-logger-be';
import { plainToClass } from 'class-transformer';
import DataTransferJobsMetadataModel, { DataTransferJobsMetadataModelToken } from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/model/mongoose/data-transfer-jobs-metadata';
import DataTransferJobsMetadata from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/data-transfer-jobs-metadata';
import { Document } from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/mongodb/document';

import { mergeMemo } from '../../core/utils/lodash';

import { PartialDataTransferJobsMetadata } from './data-transfer-jobs-metadata.controller';

@Injectable()
export class DataTransferJobsMetadataDal {
  private readonly logger = new Logger(DataTransferJobsMetadataDal.name);

  constructor(@Inject(DataTransferJobsMetadataModelToken) private readonly model: Model<Document<DataTransferJobsMetadata>>) {
  }

  findAll(filter: FilterQuery<Document<DataTransferJobsMetadata>> = {}, sort: string | any = {}, options: any = {}) {
    return this.model
      .find(filter, {}, options)
      .sort(sort)
      .exec();
  }

  findOne(filter: FilterQuery<Document<DataTransferJobsMetadata>> = {}, projection: any = {}, options: any = {}) {
    return this.model.findOne(filter, projection, options).exec();
  }

  findById(id: any, projection: any = {}, options: any = {}) {
    return this.model.findById(id, projection, options).exec();
  }

  insert(docs: any[]) {
    return this.model.insertMany(plainToClass(DataTransferJobsMetadataModel, docs));
  }

  async updateOne(_id: any, docs: UpdateQuery<PartialDataTransferJobsMetadata>) {
    const prev = await this.findById(_id, { _id: 0 }, { lean: true });
    const updated = mergeMemo(prev, docs);
    await this.model.updateOne({ _id: ObjectId.createFromHexString(_id) }, updated);
    return updated;
  }

  findByIdAndDelete(id: any) {
    return this.model.findByIdAndDelete(id);
  }
}
