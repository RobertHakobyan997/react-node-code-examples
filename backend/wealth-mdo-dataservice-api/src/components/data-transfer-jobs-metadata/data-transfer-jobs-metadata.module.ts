import { Module } from '@nestjs/common';
import DataTransferJobsMetadataModel, { DataTransferJobsMetadataModelToken } from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/model/mongoose/data-transfer-jobs-metadata';

import { HolidayCalendarsModule } from '../holiday-calendars/holiday-calendars.module';
import { DataTransferJobsStatusModule } from '../data-transfer-jobs-status/data-transfer-jobs-status.module';
import { DatabaseModule } from '../../core/database/mongodb/database.module';

import { DataTransferJobsMetadataController } from './data-transfer-jobs-metadata.controller';
import { DataTransferJobsMetadataService } from './data-transfer-jobs-metadata.service';
import { DataTransferJobsMetadataDal } from './data-transfer-jobs-metadata.dal';

@Module({
  imports: [
    DatabaseModule,
    HolidayCalendarsModule,
    DataTransferJobsStatusModule
  ],
  controllers: [ DataTransferJobsMetadataController ],
  providers  : [
    { provide: DataTransferJobsMetadataModelToken, useFactory: () => DataTransferJobsMetadataModel },
    DataTransferJobsMetadataService,
    DataTransferJobsMetadataDal
  ],
  exports: [ DataTransferJobsMetadataService ],
})
export class DataTransferJobsMetadataModule {
}
