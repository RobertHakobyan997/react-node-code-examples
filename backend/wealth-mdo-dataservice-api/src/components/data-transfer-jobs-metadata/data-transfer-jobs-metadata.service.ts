import { from, of, pipe, zip } from 'rxjs';
import { BadRequestException, Injectable } from '@nestjs/common';
import { map, switchMap } from 'rxjs/operators';
import DataTransferJobsMetadata from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/data-transfer-jobs-metadata';
import { DateTime } from 'luxon';
import { isNil } from 'lodash';
import DataTransferJobsStatusState from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/data-transfer-jobs-status-state';
import ScheduleDay from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/schedule-day';

import { HolidayCalendarsService } from '../holiday-calendars/holiday-calendars.service';
import { DataTransferJobsStatusService } from '../data-transfer-jobs-status/data-transfer-jobs-status.service';

import { PartialDataTransferJobsMetadata } from './data-transfer-jobs-metadata.controller';
import { getCurrentDayBasedOnPeriodDates, getRecordWithSuppliers, shouldResetStateRepeatCount } from './pure/data-transfer-jobs-metadata.pure';
import { DataTransferJobsMetadataDal } from './data-transfer-jobs-metadata.dal';

export class DataTransferJobsMetadataAndState extends DataTransferJobsMetadata {
  state: DataTransferJobsStatusState;
}

export class ScheduleDayWithPeriodDates extends ScheduleDay {
  state: DataTransferJobsStatusState;
  periodStartDate: DateTime;
  periodEndDate: DateTime;
}

@Injectable()
export class DataTransferJobsMetadataService {
  constructor(
    private readonly dal: DataTransferJobsMetadataDal,
    private readonly holidayCalendarsService: HolidayCalendarsService,
    private readonly dtJobStatusService: DataTransferJobsStatusService,
  ) {
  }

  findAll() {
    return this.dal.findAll();
  }

  findById(id: string) {
    return this.dal.findById(id);
  }

  async updateOne(id: string, dataTransferJobsMetadata: PartialDataTransferJobsMetadata) {
    const existingMetadata = await this.dal.findById(id, {}, { lean: true });
    if (!existingMetadata) throw new BadRequestException(`Data Transfer Metadata with id ${id} does not exist in the database`);
    return this.dal.updateOne(id, dataTransferJobsMetadata);
  }

  create(body: DataTransferJobsMetadata[]) {
    return this.dal.insert(body);
  }

  delete(id: string) {
    return this.dal.findByIdAndDelete(id);
  }

  getJobKeys() {
    return this.getSchedule()
      .pipe(
        map(schedule => schedule.map(data => data.dataTransferJobKey).sort())
      );
  }

  getSchedule() {
    return from(this.getScheduleData())
      .pipe(getRecordWithSuppliers(), this.resetTotalSuccessCount());
  }

  resetTotalSuccessCount() {
    return pipe(
      switchMap((jobs: DataTransferJobsMetadataAndState[]) => {
        const jobsForUpdate = jobs.filter(({ state, schedule }) => {
          if (isNil(state)) {
            return false;
          }

          const currentDay = getCurrentDayBasedOnPeriodDates(schedule.expectedDays as ScheduleDayWithPeriodDates[]);
          if (!currentDay) {
            return false;
          }
          return shouldResetStateRepeatCount(state, currentDay);
        });

        const updatedStatePromises = jobsForUpdate.map(
          item =>
            this.dtJobStatusService.updateJobMetadataStatuses(item._id.toString(), { state: { ...item.state, totalSuccessCount: 0 } })
        );
        const updatedJobs = jobs.map(
          item => jobsForUpdate.some(({ _id }) => _id.equals(item._id))
            ? { ...item, state: { ...item.state, totalSuccessCount: 0 } }
            : item
        );
        return zip(of(updatedJobs), from(Promise.all(updatedStatePromises)));
      }),
      map(([ jobs ]) => jobs)
    );
  }

  getScheduleData() {
    return Promise.all([
      this.dal.findAll({ isEnabled: true }, {}, { lean: true }),
      this.holidayCalendarsService.findAll({}, {}, { lean: true }),
      this.dtJobStatusService.findAll({}, {}, { lean: true }),
    ]);
  }

  async findByJobKey(dataTransferJobKey: string) {
    const job = await this.dal.findOne({ dataTransferJobKey }, {}, { lean: true });
    if (job) {
      const status = await this.dtJobStatusService.getJobMetadataStatuses(job._id.toString());
      return status ? { ...job, state: status.state } : job;
    }

    return null;
  }
}
