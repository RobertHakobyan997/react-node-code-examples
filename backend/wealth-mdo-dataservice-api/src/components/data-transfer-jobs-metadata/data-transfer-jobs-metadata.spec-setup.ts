import { Test, TestingModule } from '@nestjs/testing';
import { DataTransferJobsMetadataModelToken } from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/model/mongoose/data-transfer-jobs-metadata';
import { HolidayCalendarsModelToken } from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/model/mongoose/holiday-calendars';
import { DataTransferJobsStatusModelToken } from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/model/mongoose/data-transfer-jobs-status';
import { RegionsModelToken } from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/model/mongoose/regions';
import { CountriesModelToken } from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/model/mongoose/countries';

import { mockMongooseModel } from '../../../test/mocks/mock-mongoose.model';
import { HolidayCalendarsService } from '../holiday-calendars/holiday-calendars.service';
import { HolidayCalendarsDal } from '../holiday-calendars/holiday-calendars.dal';
import { DataTransferJobsStatusService } from '../data-transfer-jobs-status/data-transfer-jobs-status.service';
import { DataTransferJobsStatusDal } from '../data-transfer-jobs-status/data-transfer-jobs-status.dal';
import { RegionsService } from '../regions/regions.service';
import { RegionsDal } from '../regions/regions.dal';
import { CountriesService } from '../countries/countries.service';
import { CountriesDal } from '../countries/countries.dal';

import { DataTransferJobsMetadataService } from './data-transfer-jobs-metadata.service';
import { DataTransferJobsMetadataDal } from './data-transfer-jobs-metadata.dal';

export const assert = async () => {
  const module: TestingModule = await Test.createTestingModule({
    providers: [
      DataTransferJobsMetadataService,
      DataTransferJobsMetadataDal,
      DataTransferJobsStatusService,
      DataTransferJobsStatusDal,
      HolidayCalendarsService,
      HolidayCalendarsDal,
      RegionsService,
      RegionsDal,
      CountriesService,
      CountriesDal,
      { provide: HolidayCalendarsModelToken, useValue: mockMongooseModel },
      { provide: RegionsModelToken, useValue: mockMongooseModel },
      { provide: CountriesModelToken, useValue: mockMongooseModel },
      { provide: DataTransferJobsMetadataModelToken, useValue: mockMongooseModel },
      { provide: DataTransferJobsStatusModelToken, useValue: mockMongooseModel },
    ],
  }).compile();

  return module.get<DataTransferJobsMetadataService>(DataTransferJobsMetadataService);
};
