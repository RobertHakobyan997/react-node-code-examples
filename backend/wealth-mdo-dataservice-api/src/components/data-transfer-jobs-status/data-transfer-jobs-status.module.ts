import { Module } from '@nestjs/common';
import DataTransferJobsStatusModel, { DataTransferJobsStatusModelToken } from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/model/mongoose/data-transfer-jobs-status';

import { DatabaseModule } from '../../core/database/mongodb/database.module';

import { DataTransferJobsStatusController } from './data-transfer-jobs-status.controller';
import { DataTransferJobsStatusService } from './data-transfer-jobs-status.service';
import { DataTransferJobsStatusDal } from './data-transfer-jobs-status.dal';

@Module({
  imports    : [ DatabaseModule ],
  controllers: [ DataTransferJobsStatusController ],
  providers  : [
    { provide: DataTransferJobsStatusModelToken, useFactory: () => DataTransferJobsStatusModel },
    DataTransferJobsStatusService,
    DataTransferJobsStatusDal
  ],
  exports: [ DataTransferJobsStatusService ],
})
export class DataTransferJobsStatusModule {
}
