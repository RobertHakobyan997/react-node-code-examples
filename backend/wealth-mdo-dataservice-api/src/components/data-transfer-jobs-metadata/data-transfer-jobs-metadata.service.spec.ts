import DataTransferJobsMetadata from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/data-transfer-jobs-metadata';
import { Document } from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/mongodb/document';
import { plainToClass } from 'class-transformer';
import DataTransferJobsStatus from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/data-transfer-jobs-status';
import { of } from 'rxjs';
import { DateTime, Settings } from 'luxon';

import {
  dataTransferJobsMetadata,
  dataTransferJobsMetadataRaw
} from '../../../test/data/data-transfer-jobs/data-transfer-jobs-metadata.const';
import { dataTransferJobsStatus } from '../../../test/data/data-transfer-jobs/data-transfer-jobs-status.const';
import { testHolidayCalendars } from '../../../test/holiday-calendars/holiday-calendars.const';

import { DataTransferJobsMetadataAndState, DataTransferJobsMetadataService } from './data-transfer-jobs-metadata.service';
import { assert } from './data-transfer-jobs-metadata.spec-setup';

describe('ProcessingStatusesService', () => {
  let service: DataTransferJobsMetadataService;

  beforeEach(async () => {
    service = await assert();
  });

  it('plainToClass', () => {
    const mapped = plainToClass(DataTransferJobsMetadata, JSON.parse(JSON.stringify(dataTransferJobsMetadataRaw)));
    // ToDo: fix structure and entity in common
    // expect(mapped._id).toBeInstanceOf(Types.ObjectId);
    expect(mapped).toMatchSnapshot();
  });

  describe('findAll()', () => {
    it('should return all data transfer jobs metadata', async () => {
      jest.spyOn(service['dal'], 'findAll').mockResolvedValue(dataTransferJobsMetadata as Document<DataTransferJobsMetadata>[]);
      const res = await service.findAll();
      expect(res).toEqual(dataTransferJobsMetadata);
      expect(res.length).toBe(1);
    });
  });

  describe('findById()', () => {
    it('should return data transfer job metadata by id', async () => {
      jest.spyOn(service['dal'], 'findById').mockResolvedValue(dataTransferJobsMetadata[0] as Document<DataTransferJobsMetadata>);
      const res = await service.findById(dataTransferJobsMetadata[0]._id.toString());
      expect(res).toEqual(dataTransferJobsMetadata[0]);
    });
  });

  describe('create()', () => {
    it('should create and return data transfer job metadata', async () => {
      jest.spyOn(service['dal'], 'insert').mockImplementation((data: DataTransferJobsMetadata[]) => data as any);
      const res = await service.create(dataTransferJobsMetadata);
      expect(res).toEqual(dataTransferJobsMetadata);
      expect(service['dal'].insert).toHaveBeenCalledWith(dataTransferJobsMetadata);
    });
  });

  describe('delete()', () => {
    it('should delete data transfer job metadata and return result of the operation', async () => {
      jest.spyOn(service['dal'], 'findByIdAndDelete').mockResolvedValue(dataTransferJobsMetadata[0] as Document<DataTransferJobsMetadata>);
      await service.delete(dataTransferJobsMetadata[0]._id.toString());
      expect(service['dal'].findByIdAndDelete).toHaveBeenCalledWith(dataTransferJobsMetadata[0]._id.toString());
    });
  });

  describe('findByJobKey()', () => {
    it('should return data transfer job metadata with state by job key', async () => {
      jest.spyOn(service['dal'], 'findOne').mockResolvedValue(dataTransferJobsMetadata[0] as Document<DataTransferJobsMetadata>);
      jest.spyOn(service['dtJobStatusService'], 'getJobMetadataStatuses')
        .mockResolvedValue(dataTransferJobsStatus[0] as Document<DataTransferJobsStatus>);
      const res = await service.findByJobKey(dataTransferJobsMetadata[0].dataTransferJobKey);
      expect(res).toEqual({ ...dataTransferJobsMetadata[0], state: dataTransferJobsStatus[0].state });
    });

    it('should return data transfer job metadata without state by job key', async () => {
      jest.spyOn(service['dal'], 'findOne').mockResolvedValue(dataTransferJobsMetadata[0] as Document<DataTransferJobsMetadata>);
      jest.spyOn(service['dtJobStatusService'], 'getJobMetadataStatuses')
        .mockResolvedValue(null);
      const res = await service.findByJobKey(dataTransferJobsMetadata[0].dataTransferJobKey);
      expect(res).toEqual(dataTransferJobsMetadata[0]);
    });
  });

  describe('getScheduleData()', () => {
    it('should return data transfer job metadata, statuses and holiday calendars', async () => {
      jest.spyOn(service['dal'], 'findAll').mockResolvedValue(dataTransferJobsMetadata as any);
      jest.spyOn(service['holidayCalendarsService'], 'findAll')
        .mockResolvedValue(testHolidayCalendars as any);
      jest.spyOn(service['dtJobStatusService'], 'findAll')
        .mockResolvedValue(dataTransferJobsStatus as any);
      const res = await service.getScheduleData();
      expect(res).toEqual([ dataTransferJobsMetadata, testHolidayCalendars, dataTransferJobsStatus ]);
    });
  });

  describe('resetTotalSuccessCount()', () => {
    it('should return metadata without state update', async () => {
      const job = { schedule: dataTransferJobsMetadata[0].schedule, state: dataTransferJobsStatus[0].state };
      const data = [ job ] as DataTransferJobsMetadataAndState[];
      jest.spyOn(service['dtJobStatusService'], 'updateJobMetadataStatuses').mockResolvedValue('test');
      const res = await of(data).pipe(service.resetTotalSuccessCount()).toPromise();
      expect(res).toEqual(data);
      expect(service['dtJobStatusService'].updateJobMetadataStatuses).not.toHaveBeenCalled();
    });

    it('should return metadata with state update', async () => {
      const expectedDays = dataTransferJobsMetadata[0].schedule.expectedDays.map(item => ({
        ...item,
        periodStartDate: DateTime.fromISO('2021-08-06T05:00:30.437Z'),
        periodEndDate  : DateTime.fromISO('2021-08-06T23:59:59.437Z')
      }));
      const scheduleWithPeriod = {
        ...dataTransferJobsMetadata[0].schedule,
        expectedDays
      };
      const state = {
        ...dataTransferJobsStatus[0].state,
        lastSuccessfulProcessTime: '2021-08-05T17:20:30.437Z',
      };
      const job = { _id: dataTransferJobsMetadata[0]._id, schedule: scheduleWithPeriod, state };
      const resultState = { ...state, totalSuccessCount: 0 };
      const resultJob = { ...job, state: resultState };
      const data = [ job ] as unknown as DataTransferJobsMetadataAndState[];

      Settings.now = () => new Date('2021-08-06T15:00:00.000Z').valueOf();
      jest.spyOn(service['dtJobStatusService'], 'updateJobMetadataStatuses').mockResolvedValue('test');
      const res = await of(data).pipe(service.resetTotalSuccessCount()).toPromise();
      expect(res).toEqual([ resultJob ]);
      expect(service['dtJobStatusService'].updateJobMetadataStatuses).toHaveBeenCalledTimes(1);
    });
  });
});
