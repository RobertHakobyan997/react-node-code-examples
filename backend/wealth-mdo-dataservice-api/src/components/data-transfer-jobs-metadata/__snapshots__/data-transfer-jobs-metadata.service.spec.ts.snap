// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProcessingStatusesService plainToClass 1`] = `
DataTransferJobsMetadata {
  "_id": "611f68a99f217cce3a4d91f8",
  "data": Array [
    DataTransferJobsMetadataData {
      "_id": "611f6d0df72945a1fb538c0d",
      "accountName": "EaglePaceAPI",
      "childObjectName": "genericEntity",
      "columnName": "upd_Date",
      "dataHeader": "entityTransactionMessage",
      "dbName": "rulesdbo",
      "eagleApiEndPointName": "entities",
      "eagleIdentifiers": EagleIdentifiers {
        "accountIdType": "NRS_CUSTODY_FUND",
        "securityIdType": "NRS_CUSTODY",
      },
      "eaglePaceURL": "https://mgi-d003-star.eagleaccess.com/eagle/v2",
      "ftpAccountName": "EaglePaceSFTP",
      "ftpOutputPath": "ftp://intranetftp.mercer.com/public/wealth/MDO/output/",
      "ftpPath": "ftp://intranetftp.mercer.com/public/wealth/MDO/",
      "historyOnlyFlag": "N",
      "objectType": "Entities",
      "parentObjectName": "entityTransaction",
      "tableName": "entity",
      "topicName": "P500_EAGLE2KF_ENTITIES",
    },
  ],
  "dataTransferJobKey": "eagle2Kafka_Entities",
  "isEnabled": true,
  "notification": Array [
    Notification {
      "escalationLevel": 1,
      "notificationIds": Array [
        "mikhail.krastelev@mmc.com",
        "tagaibek.tolubaev@mmc.com",
        "kiran.dhage@mercer.com",
        "prabhukumar.a@mercer.com",
        "saran.thiruchabai@mmc.com",
        "nemi.nabiraj@mmc.com",
      ],
      "notificationKey": "validation",
      "notificationMethod": "email",
    },
    Notification {
      "escalationLevel": 1,
      "notificationIds": Array [
        "mikhail.krastelev@emergn.com",
        "tagaibek.tolubaev@emergn.com",
        "kiran.dhage@mercer.com",
        "prabhukumar.a@mercer.com",
        "saran.thiruchabai@mmc.com",
        "nemi.nabiraj@mmc.com",
      ],
      "notificationKey": "alerts",
      "notificationMethod": "email",
    },
  ],
  "region": "unitedStates",
  "schedule": DataTransferJobsSchedule {
    "endDate": "2023-09-01",
    "expectedDays": Array [
      DataTransferJobsScheduleDay {
        "dayOfTheWeek": "AllWorkDays",
        "endOfLate": ScheduleTime {
          "time": "17:00",
        },
        "endOfOnSchedule": ScheduleTime {
          "time": "13:00",
        },
        "endOfOnTime": ScheduleTime {
          "time": "16:00",
        },
      },
    ],
    "frequency": "Daily",
    "holidayCalendarId": "609d488d3413c47110d7f91f",
    "repeatDuration": "1",
    "startDate": "2021-09-01",
    "startTime": "00:01",
    "timeZone": "America/Chicago",
  },
  "snaplogicTasks": Array [
    Snaplogic {
      "camundaTopicName": "callSnaplogicDataTransferTask",
      "taskUrl": "TASK_TRGD_MDO_Eagle2Kafka_PushData",
    },
  ],
  "workflowProcessKey": "",
  "workflowProcessName": "DataTransfer_Default_Workflow",
}
`;
